src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/SimpleFileServerImpl.java

```diff
@@ -0,0 +1,203 @@
+
+    private SimpleFileServerImpl() { throw new AssertionError(); }
+
+    /**
+     * Starts a simple HTTP file server created on a directory.
+     *
+     * @param  writer the writer to which output should be written
+     * @param  args the command line options
+     * @throws NullPointerException if any of the arguments are {@code null},
+     *         or if there are any {@code null} values in the {@code args} array
+     * @return startup status code
+     */
+    static int start(PrintWriter writer, String[] args) {
+        Objects.requireNonNull(args);
+        for (var arg : args) {
+            Objects.requireNonNull(arg);
+        }
+        Out out = new Out(writer);
+
+        InetAddress addr = DEFAULT_ADDR;
+        int port = DEFAULT_PORT;
+        Path root = DEFAULT_ROOT;
+        OutputLevel outputLevel = DEFAULT_OUTPUT_LEVEL;
+
+        // parse options
+        Iterator<String> options = Arrays.asList(args).iterator();
+        String option = null;
+        String optionArg = null;
+        try {
+            while (options.hasNext()) {
+                option = options.next();
+                switch (option) {
+                    case "-h", "-?", "--help" -> {
+                        out.showHelp();
+                        return Startup.OK.statusCode;
+                    }
+                    case "-b", "--bind-address" ->
+                        addr = InetAddress.getByName(optionArg = options.next());
+                    case "-d", "--directory" ->
+                        root = Path.of(optionArg = options.next());
+                    case "-o", "--output" ->
+                        outputLevel = Enum.valueOf(OutputLevel.class,
+                                (optionArg = options.next()).toUpperCase(Locale.ROOT));
+                    case "-p", "--port" ->
+                        port = Integer.parseInt(optionArg = options.next());
+                    default -> throw new AssertionError();
+                }
+            }
+        } catch (AssertionError ae) {
+            out.reportError(ResourceBundleHelper.getMessage("err.unknown.option", option));
+            out.showUsage();
+            return Startup.CMDERR.statusCode;
+        } catch (NoSuchElementException nsee) {
+            out.reportError(ResourceBundleHelper.getMessage("err.missing.arg", option));
+            return Startup.CMDERR.statusCode;
+        } catch (Exception e) {
+            out.reportError(ResourceBundleHelper.getMessage("err.invalid.arg", option, optionArg));
+            e.printStackTrace(out.writer);
+            return Startup.CMDERR.statusCode;
+        } finally {
+            out.flush();
+        }
+
+        // configure and start server
+        try {
+            var socketAddr = new InetSocketAddress(addr, port);
+            var server = SimpleFileServer.createFileServer(socketAddr, root, outputLevel);
+            server.setExecutor(Executors.newSingleThreadExecutor());
```

Comment:
I think this code has the effect of creating one thread for the selector and a second one for the execution of the handlers. If we want to keep thread usage to an absolute minimum then it might be better to not set an executor. Then, the selector thread executes the handlers as well.

Reply:
I did a quick stress test of both versions (with and without executor specified) and don’t see any difference in how many requests can be handled. Probably good to keep thread usage to a minimum, I'll update that.